File: visualizer.py
from typing import List
from termcolor import colored  # install via: pip install termcolor

class Solution:
    def print_indent(self, level):
        return "  " * level

    def backtrack(self, nums: List[int], start: int, current: List[int], result: List[List[int]], depth=0):
        indent = self.print_indent(depth)
        print(indent + colored(f"Enter: start={start}, current={current}", "cyan"))

        result.append(list(current))

        for i in range(start, len(nums)):
            # Skip duplicates
            if i > start and nums[i] == nums[i - 1]:
                print(indent + colored(f"Skip duplicate: nums[{i}] = {nums[i]}", "red"))
                continue

            print(indent + colored(f"Include: nums[{i}] = {nums[i]}", "green"))
            current.append(nums[i])

            self.backtrack(nums, i + 1, current, result, depth + 1)

            print(indent + colored(f"Backtrack: remove nums[{i}] = {nums[i]}", "yellow"))
            current.pop()

        print(indent + colored(f"Exit: start={start}, current={current}", "cyan"))

    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()  # Sort to handle duplicates
        result = []
        self.backtrack(nums, 0, [], result)
        return result


# Example usage
if __name__ == "__main__":
    nums = [1, 2, 2]
    sol = Solution()
    print("\n--- Backtracking Process ---\n")
    result = sol.subsetsWithDup(nums)

    print("\n--- All Unique Subsets ---\n")
    for subset in result:
        print(subset)

File: recurssion/visualizer.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    void printIndent(int level) {
        for (int i = 0; i < level; ++i) cout << "  ";
    }

    void backtrack(vector<int>& nums, int start, vector<int>& current, vector<vector<int>>& result, int depth = 0) {
        printIndent(depth);
        cout << "Enter: start=" << start << ", current={";
        for (int num : current) cout << num << " ";
        cout << "}" << endl;

        result.push_back(current); // Save current subset

        for (int i = start; i < nums.size(); ++i) {
            if (i > start && nums[i] == nums[i - 1]) {
                printIndent(depth);
                cout << "Skip duplicate: nums[" << i << "]=" << nums[i] << endl;
                continue;
            }

            current.push_back(nums[i]);
            printIndent(depth);
            cout << "Include: nums[" << i << "]=" << nums[i] << endl;

            backtrack(nums, i + 1, current, result, depth + 1);

            current.pop_back();
            printIndent(depth);
            cout << "Backtrack: removed nums[" << i << "]=" << nums[i] << endl;
        }

        printIndent(depth);
        cout << "Exit: start=" << start << ", current={";
        for (int num : current) cout << num << " ";
        cout << "}" << endl;
    }

    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> result;
        vector<int> current;
        backtrack(nums, 0, current, result);
        return result;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {1, 2, 2};
    vector<vector<int>> subsets = sol.subsetsWithDup(nums);

    cout << "\nAll Unique Subsets:\n";
    for (auto& subset : subsets) {
        cout << "{ ";
        for (int num : subset) cout << num << " ";
        cout << "}\n";
    }

    return 0;
}
